python/foglamp/plugins/south/sinusoid/sinusoid.py

```diff
@@ -166,35 +168,40 @@ def generate_data():
 
     async def save_data():
         try:
+            time_time = time.time
+            start = time_time()
+            recs = int(handle['dataPointsPerSec']['value'])
+            period = round(1.0 / recs, len(str(recs))+1)
             while True:
-                time_stamp = utils.local_timestamp()
-                data = {
-                    'asset': handle['assetName']['value'],
-                    'timestamp': time_stamp,
-                    'key': str(uuid.uuid4()),
-                    'readings': {
-                        "sinusoid": next(generate_data())
+                now = time.time()
+                if (now - start) > period:
+                    start += period
+                    time_stamp = utils.local_timestamp()
+                    data = {
+                        'asset': handle['assetName']['value'],
+                        'timestamp': time_stamp,
+                        'key': str(uuid.uuid4()),
+                        'readings': {
+                            "sinusoid": next(generate_data())
+                        }
                     }
-                }
-
-                await Ingest.add_readings(asset='{}'.format(data['asset']),
-                                          timestamp=data['timestamp'], key=data['key'],
-                                          readings=data['readings'])
-
-                try:
-                    await asyncio.sleep(1 / (float(handle['dataPointsPerSec']['value'])))
-                except ZeroDivisionError:
-                    _LOGGER.warning('Data points per second must be greater than 0, defaulting to 1')
-                    handle['dataPointsPerSec']['value'] = '1'
-                    await asyncio.sleep(1)
+                    await Ingest.add_readings(asset='{}'.format(data['asset']),
+                                              timestamp=data['timestamp'], key=data['key'],
+                                              readings=data['readings'])
         except asyncio.CancelledError:
             pass
         except (Exception, RuntimeError) as ex:
             _LOGGER.exception("Sinusoid exception: {}".format(str(ex)))
             raise exceptions.DataRetrievalError(ex)
 
-    _task = asyncio.ensure_future(save_data())
+    def run_task(loop):
+        global _task
+        asyncio.set_event_loop(loop)
+        _task = asyncio.ensure_future(save_data(), loop=loop)
 
+    new_loop = asyncio.get_event_loop()
+    t = Thread(target=run_task, args=(new_loop,))
+    t.start()
```

Comment:
could you please share the results of shutdwn and reconfigure?

Reply:
done